### PR TITLE
Chart annotations

### LINKED ISSUE #2191: Chart: add support for Line Annotations plugin
**Is your feature request related to a problem? Please describe.**
 I am trying to do https://www.chartjs.org/chartjs-plugin-annotation/guide/types/line.html#configuration
image
how can i do it

### PR DESCRIPTION
**Pull request by stsrki**
Closes #2191

### CODE REVIEW

File: Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import { getChart } from "../Blazorise.Charts/charts.js?v=1.2.2.0";
+import { parseFunction } from "../Blazorise.Charts/utilities.js?v=1.2.2.0";
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js:2]: we already have an open issue for v1.3 for this, but we need to update to v1.3. Just friendly reminder.

File: Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+            chart.options.plugins.annotation = adjustOptions(options);
+        }
+
+        chart.update();
+    }
+}
+
+function adjustOptions(options) {
+    return { annotations: options };
+}
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js:22]: do we really need this function? I think it's more confusing to read then just assigning the object to annotation?

File: Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+function compileDatasetsOptionsCallbacks(options) {
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js:24]: Is this used?

File: Source/Extensions/Blazorise.Charts/ChartColor.cs [hunk 4]
```diff
@@ -40,8 +41,123 @@ public ChartColor( float red, float green, float blue, float alpha )
 
     #endregion
 
+    #region Operators
+
+    /// <summary>
+    /// Implicitly converts from string value to <see cref="ChartColor"/>.
+    /// </summary>
+    /// <param name="value">String value.</param>
+    public static implicit operator ChartColor( string value )
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Charts/ChartColor.cs:50]: Does not support, "red, blue,etc..." I'm guessing?

File: Source/Extensions/Blazorise.Charts/ChartColor.cs [hunk 5]
```diff
@@ -40,8 +41,123 @@ public ChartColor( float red, float green, float blue, float alpha )
...
+        var sb = new StringBuilder();
+        var result = IsHexDigit.Matches( input );
+
+        foreach ( System.Text.RegularExpressions.Match item in result )
+        {
+            sb.Append( item.Value );
+        }
+
+        return sb.ToString();
+    }
```
Comment (David-Moreira) [inline Source/Extensions/Blazorise.Charts/ChartColor.cs:85]: Isn't there somewhere else with this code? Can it be shared?

**Review by David-Moreira — APPROVED** [review]
LGTM. Maybe in the docs we can show the other example thats in the demo too? The one where you can label different parts of the chart? 
Users might not know that's possible.

File: Source/Extensions/Blazorise.Charts.Annotation/ChartAnnotationOptionsConverter.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    public override ChartAnnotationOptions Read( ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options )
+    {
+        throw new NotImplementedException();
+    }
+
+    public override void Write( Utf8JsonWriter writer, ChartAnnotationOptions value, JsonSerializerOptions options )
+    {
+        JsonSerializer.Serialize( writer, (object)value, value.GetType() );
+    }
+}
```
[the remark this answers is not in the document]
Comment (stsrki) [reply Source/Extensions/Blazorise.Charts.Annotation/ChartAnnotationOptionsConverter.cs:20]: I prefer it separate for now.

File: Source/Extensions/Blazorise.Charts/ChartColor.cs (hunk 4 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Charts/ChartColor.cs:50]: Nop.

File: Source/Extensions/Blazorise.Charts/ChartColor.cs (hunk 5 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Charts/ChartColor.cs:85]: There is in ThemeGenerator. Didn't want to make it public so and was easy to just copy/paste.

File: Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js (hunk 2 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js:22]: Well, it adjusts the options for the plugin. I guess its fine.

Comment (David-Moreira) [reply Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js]: I just see a regular object assignment here. The naming of the function leads me to think this actually does something with the options.
Anyway I'm fine as is.

Comment (stsrki) [reply Source/Extensions/Blazorise.Charts.Annotation/wwwroot/chart.annotation.js]: In this case, the annotations options are not complicated. But I wanted to make it similar to another extension.

File: Demos/Blazorise.Demo/Pages/Tests/ChartsAnnotationsPage.razor.cs [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+                BackgroundColor = "rgba(159, 168, 218, 0.2)",
+                BorderWidth = 0,
+                XMin = 8.5,
+                Label = new()
+                {
+                    DrawTime = "afterDraw",
+                    Display = true,
+                    Content = "Fourth quarter",
+                    Position = new { x = "center", y = "start" }
+                },
```
[the remark this answers is not in the document]
Comment (stsrki) [reply Demos/Blazorise.Demo/Pages/Tests/ChartsAnnotationsPage.razor.cs:181]: Fixed demo, and added it to the docs.

**Pull request merged**